### PR TITLE
The new ID implementation

### PR DESCRIPTION
**Pull request by arpitbbhayani**
Benchmark results. We can tune the widths to accommodate the buffer size and cap the throughput and tolerance.

```
goos: linux
goarch: amd64
pkg: github.com/dicedb/dice/core
cpu: AMD Ryzen 9 7950X 16-Core Processor            
BenchmarkNextID-32              1000000000               4.515 ns/op           0 B/op          0 allocs/op
PASS
ok      github.com/dicedb/dice/core     6.838s
```

Approach

image0

### CODE REVIEW

File: core/id.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+var mu sync.Mutex
+var cycle uint32 = 0
+var counter uint32 = 0
+var turn uint32 = 0
+
+var totalBits int = 32
+var turnBits int = 8
+var counterBits int = totalBits - turnBits
+
+func ExpandID(id uint32) uint64 {
```
[the remark this answers is not in the document]
Comment (princejha95) [reply core/id.go:16]: Also a liitle bit of documentation on what purpose the variables serve will be very helpful in understanding the flow.

File: core/id_test.go [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package core
+
+import (
+	"testing"
+)
+
+func BenchmarkNextID(b *testing.B) {
```
Comment (princejha95) [inline core/id_test.go:7]: Documenting the results of the testing will be helpful as tomorrow if the performance of the logic which is getting tested gets degraded, we will atleast know about the previous benchmark data and then move in the direction of fixing the issue.

File: core/id.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (arpitbbhayani) [reply core/id.go:20]: @princejha95 PR is in draft state. subject to a lot of changes.

File: core/id.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+// TODO: Persisting the cycle on disk
+// and reloading it when we start the server
+func NextID() uint32 {
+	mu.Lock()
+	defer mu.Unlock()
+	counter = (counter + 1) & ((1 << counterBits) - 1)
+	if counter == 0 {
+		cycle++
+		turn = (turn + 1) & ((1 << turnBits) - 1)
```
Comment (uds5501) [inline core/id.go:34]: Q : If the turn reaches 0 again (wraps around after `turn=16`), it'll re-write the cycle. Is it expected that transaction IDs of earlier `turn=16` will no be present by the time this cycle comes through?

Comment (uds5501) [reply core/id.go:34]: Ah nvm, saw your buffer & disambiguate point, it's indeed not expected to be present.

**Pull request merged**